src/core/shared_memory_manager.cc

```diff
@@ -0,0 +1,244 @@
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+#include "src/core/shared_memory_manager.h"
+
+#include <fcntl.h>
+#include <sys/mman.h>
+#include <unistd.h>
+#include <algorithm>
+#include <deque>
+#include <exception>
+#include <future>
+#include "src/core/constants.h"
+#include "src/core/logging.h"
+#include "src/core/server_status.h"
+
+namespace nvidia { namespace inferenceserver {
+
+Status
+open_shm_region(const std::string& shm_key, int* shm_fd)
+{
+  // get shared memory region descriptor
+  *shm_fd = shm_open(shm_key.c_str(), O_RDONLY, S_IRUSR | S_IWUSR);
+  if (*shm_fd == -1) {
+    return Status(
+        RequestStatusCode::INTERNAL, "Unable to get shared memory descriptor");
+  }
+
+  return Status::Success;
+}
+
+Status
+get_shm_addr(
+    const int shm_fd, const size_t offset, const size_t byte_size,
+    void** shm_addr)
+{
+  // map shared memory to process address space
+  *shm_addr = mmap(NULL, byte_size, PROT_WRITE, MAP_SHARED, shm_fd, offset);
+  if (*shm_addr == MAP_FAILED) {
+    return Status(
+        RequestStatusCode::INTERNAL, "Unable to process address space");
+  }
+
+  return Status::Success;
+}
+
+Status
+shm_close(int shm_fd)
+{
+  int tmp = close(shm_fd);
+  if (tmp == -1) {
+    return Status(
+        RequestStatusCode::INTERNAL, "Unable to close shared memory region");
+  }
+
+  return Status::Success;
+}
+
+Status
+mmap_cleanup(void* shm_addr, size_t byte_size)
+{
+  int tmp_fd = munmap(shm_addr, byte_size);
+  if (tmp_fd == -1) {
+    return Status(
+        RequestStatusCode::INTERNAL, "Unable to munmap shared memory region");
+  }
+
+  return Status::Success;
+}
+
+Status
+SharedMemoryManager::RegisterSharedMemory(
+    const std::string& name, const std::string& shm_key, const size_t offset,
+    const size_t byte_size)
+{
+  LOG_VERBOSE(1) << "Register() shared memory region: '" << name << "'";
+  Status status = Status::Success;
+
+  // check if key is in shared_memory_map_ then remove
+  if (shared_memory_map_.find(name) != shared_memory_map_.end()) {
```

Comment:
This can be dangerous if client makes an mistake are reuses name for 2 different regions. Instead we should return an error.